src/main/java/seedu/address/model/VersionedAddressBook.java

```diff
@@ -0,0 +1,156 @@
+package seedu.address.model;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+import javafx.collections.ObservableList;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.person.Person;
+
+/**
+ * Represents the different states of the Address Book for the user.
+ */
+public class VersionedAddressBook extends AddressBook {
+    public static final String MESSAGE_NO_MORE_HISTORY = "This is the earliest version that user can retrieve";
+    public static final String MESSAGE_UNSAVED_CHANGES = "There are unsaved changes in the current state. "
+            + "Please commit or discard the changes before undoing.";
+    private final ArrayList<AddressBook> addressBookStateList;
+    private int currentStatePointer;
+    private AddressBook current;
+
+
+    /**
+     * Constructs an empty state list for the address book.
+     */
+    public VersionedAddressBook() {
+        addressBookStateList = new ArrayList<>();
+        currentStatePointer = -1;
+        current = new AddressBook();
+        this.commitAddressBook();
+    }
+
+    /**
+     * Constructs a state list for the address book and saves the initial copy.
+     */
+    public VersionedAddressBook(ReadOnlyAddressBook addressBook) {
+        addressBookStateList = new ArrayList<>();
+        currentStatePointer = -1;
+        current = new AddressBook(addressBook);
+        this.commitAddressBook();
+    }
+
+    /**
+     * Saves the whole address book state in the history.
+     */
+    public void commitAddressBook() {
+        currentStatePointer++;
+        addressBookStateList.subList(currentStatePointer, addressBookStateList.size()).clear();
+        addressBookStateList.add(currentStatePointer, new AddressBook(current));
+    }
+
+    /**
+     * Reverts the given AddressBook to the previous state.
+     */
+    public void undoAddressBook() throws CommandException {
+        if (currentStatePointer <= 0) {
+            throw new CommandException(MESSAGE_NO_MORE_HISTORY);
+        }
+
+        if (!current.equals(addressBookStateList.get(currentStatePointer))) {
+            throw new CommandException(MESSAGE_UNSAVED_CHANGES);
+        }
+
+        currentStatePointer--;
+        current = new AddressBook(addressBookStateList.get(currentStatePointer));
+    }
+
+    /**
+     * Discards the unsaved changes in the current state.
+     */
+    public void discardUnsavedChanges() {
+        current = new AddressBook(addressBookStateList.get(currentStatePointer));
+    }
```

Comment:
I'm not too sure about this but perhaps the user might not need this method? Since the user would not be able to call ``setAddressBook`` method, but this is the only scenario where there would be discrepancy between current Address Book and the Address Book being pointed by the pointer.

(Again, please feel free to correct me if I have mistaken anything!

Reply:
You are right in that this method may not be directly used by the user. It is more like a convenience method for programmers. Just to clarify, the discrepancy can be caused by any data-changing methods of AB. (refer to one of my comments above)

The motivation is this scenario:
A programmer made some changes to AB and realises they are not needed. Currently there is absolutely no way to reset to the last commit (the one pointed to by the pointer), because an undo will revert to the second to last commit (discarding last). In this case this method will come to rescue.